JS/TS: insecure Helmet middleware (new query)

aegilops opened the pull request:
The [Helmet 🪖 middleware](https://helmetjs.github.io/) is used to set security-related HTTP headers in Express applications. This query finds instances where the middleware is configured with important ⚠️  security features disabled 🚫.

I've included a new data extension, specific to this query, called `requiredHelmetSecuritySetting`, which users can set to a string of extra features that should not be disabled, beyond the defaults in the query. The extension is specific to this query, rather than general purpose; I don't know of other similar queries for other frameworks, so designing something that would work well in other contexts would be speculative.

I'd like to discuss 🗣️ what numeric severity 🔢 to give this, since it's a configuration and not a direct vulnerability on its own.

aegilops commented:
I'm not sure what to do about the change note failure. Do I need to document this change in a change note, and if so, where please?

erik-krogh commented:
> Do I need to document this change in a change note, and if so, where please?

Yes. 
`ql/javascript/ql/src/change-notes/yyyy-mm-dd-name-of-your-node.md`.   

See this doc on change-notes for what to put there: https://github.com/github/codeql/blob/main/docs/change-notes.md

erik-krogh commented on javascript/ql/src/Security/CWE-693/InsecureHelmet.ql:
```diff
@@ -0,0 +1,58 @@
+
+extensible predicate requiredHelmetSecuritySetting(string name);
+
+from HelmetProperty helmetSetting, ExpressLibraries::HelmetRouteHandler helmet
```
Setting or property? 

I'm not sure which one is right, but you should be consistent. 

```suggestion
from HelmetProperty helmetProperty, ExpressLibraries::HelmetRouteHandler helmet
```

erik-krogh commented on javascript/ql/src/Security/CWE-693/InsecureHelmet.ql:
```diff
@@ -0,0 +1,58 @@
+
+import semmle.javascript.frameworks.ExpressModules
+
+class HelmetProperty extends Property {
+  ExpressLibraries::HelmetRouteHandler helmet;
+
+  HelmetProperty() {
+    helmet.(DataFlow::CallNode).getAnArgument().asExpr().(ObjectExpr).getAProperty() = this
+  }
+
+  ExpressLibraries::HelmetRouteHandler getHelmet() { result = helmet }
+
+  predicate isFalse() { this.getInit().(BooleanLiteral).getBoolValue() = false }
+
+  predicate isImportantSecuritySetting() {
+    this.getName() in ["frameguard", "contentSecurityPolicy"]
+    or
+    // read from data extensions to allow enforcing other settings
+    requiredHelmetSecuritySetting(this.getName())
+  }
+}
```
Here is a suggestion for how you can use some more modern libraries.  (Staying in the dataflow layer).  

```suggestion
class HelmetProperty extends DataFlow::Node instanceof DataFlow::PropWrite {
  ExpressLibraries::HelmetRouteHandler helmet;

  HelmetProperty() {
    this = helmet.(DataFlow::CallNode).getAnArgument().getALocalSource().getAPropertyWrite()
  }

  ExpressLibraries::HelmetRouteHandler getHelmet() { result = helmet }

  predicate isFalse() { DataFlow::PropWrite.super.getRhs().mayHaveBooleanValue(true) }

  string getName() { result = DataFlow::PropWrite.super.getPropertyName() }

  predicate isImportantSecuritySetting() {
    this.getName() in ["frameguard", "contentSecurityPolicy"]
    or
    // read from data extensions to allow enforcing other settings
    requiredHelmetSecuritySetting(this.getName())
  }
}
```

erik-krogh commented on javascript/ql/src/Security/CWE-693/InsecureHelmet.ql:
```diff
@@ -0,0 +1,58 @@
+ * Note: `frameguard` is an example: the query already enforces this setting, so it is not necessary to add it to the data extension.
+ */
+
+extensible predicate requiredHelmetSecuritySetting(string name);
```
Both on CI and on my local machine I'm getting this error: 
```
The query depends on an extensional predicate requiredHelmetSecuritySetting which has not been defined.
```

I have not worked with `extensible` predicates, so I'm not sure what the fix could be, but maybe you have to add an empty extension, just so there is some data?

aegilops replied:
Yep, good point. Sorry to waste your time by not checking.

I'd assumed if it didn't find a matching extension file it'd just carry on regardless. Looks like it needs at least one matching model; and I forgot to even test that assumption locally.

What I've done now is defined a model file in the `javascript/ql/lib/semmle/frameworks` directory, which is used to define the required security settings.

I've also added a `README.md` to help guide users on how to customize this query with a data extension of their own.

erik-krogh reviewed (commented):
Some comments for now.

aegilops commented:
I've got those updates done for you @erik-krogh, thanks so much for the review so far 🙏 

* added a change note
* switched to modern dataflow libraries you suggested
* used a default data extension, since using an external predicate always requires external data
* added a `README.md` to explain how to customize the query with data extensions
* fixed the inconsistent naming
* updated the severity to 7.0, as we discussed
* moved the examples to separate files outside the `.qhelp` file
* updated the name to `js/...` (instead of `javascript/...`), which I think is the convention

erik-krogh commented on javascript/ql/src/Security/CWE-693/InsecureHelmet.qhelp:
```diff
@@ -0,0 +1,61 @@
+        <sample src="examples/helmet_custom.js" />
+
+    </example>
+    <references>
```
I think you could add some references here to read more about CSP, XSS, and frame protections.

erik-krogh reviewed (commented):
You need to update the expected output of the test, and I think we could use some more references in the QHelp.  

But the rest looks good 👍

aegilops commented:
👋 I fixed the expected result, added more references, and fixed a bit of markup in the query help.

I just updated my branch with `main` and pushed my changes, so I hope this will be a clean run and OK to merge.

erik-krogh reviewed (approved):
Nice, this looks good to me 👍 

One below comment about `dataExtensions:`. I don't know what the right answer is 🤷 

I did some evaluations to check that the results/performance look OK. 
([nightly](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-16540-5a3328__nightly-old__code-scanning/reports), [default](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-16540-5a3328__default__code-scanning/reports) (internal links)).  

Performance was great 👍 
And the evaluations found a single new TP, which looked like the below, so that's nice.  
```JS
app.use(helmet({
  contentSecurityPolicy: false, // @TODO implement
```

aegilops commented on javascript/ql/lib/semmle/javascript/frameworks/helmet/Helmet.Required.Setting.model.yml:
```diff
@@ -0,0 +1,7 @@
+extensions:
```
It's already included by default, since it is under `frameworks` - there are other models already there, so it just gets picked up along with those that are already referenced
